### ISSUE TITLE
`builtins.trace` should print lambda location

### ISSUE THREAD

**Issue by roberth** [NixOS/nix#7145]
**Is your feature request related to a problem? Please describe.**

There's no reason why `trace` shouldn't be able to do what the repl printer can:

```
nix-repl> builtins.trace lib.showOption {}
trace: <LAMBDA>
{ }

nix-repl> lib.showOption
«lambda @ /home/user/h/nixpkgs/lib/options.nix:322:16»
```

**Describe the solution you'd like**

At least print the location like the repl does.

More ideas:
 - print closure info when debugger is enabled
 - when the `builtins.trace` argument is a thunk that points to a string interpolation, copy the thunk and evaluate the string with a flag that allows the lambda to be rendered as a source location (I don't think we have a nice ast node for this though. Add a new method to `Expr` for this kind of evaluation?)

**Describe alternatives you've considered**

Use the debugger.

**Additional context**

**Comment by 9999years** [comment]
Fixed with #9606.

```
$ nix repl nixpkgs
Welcome to Nix 2.21.0pre20240131_dirty. Type :? for help.

Loading installable 'flake:nixpkgs#'...
Added 5 variables.
nix-repl> builtins.trace lib.showOption {}
trace: «lambda showOption @ /nix/store/ildml01gk7v91fba35whklllfi343lvq-source/lib/options.nix:411:16»
{ }
```

**Comment by roberth** [comment]
Thank you @9999years!



> More ideas:
> 
> * print closure info when debugger is enabled

Better to be implemented as a repl command instead. This is would be quite a lot of info, and it's easy to get a lambda result by mistake. That'd be annoying.

> * when the `builtins.trace` argument is a thunk that points to a string interpolation, copy the thunk and evaluate the string with a flag that allows the lambda to be rendered as a source location (I don't think we have a nice ast node for this though. Add a new method to `Expr` for this kind of evaluation?)

Not sure. Idea also covered by

- #7795
